### ISSUE TITLE
Please implement batch-mode viterbi decode

### ISSUE THREAD

**Issue by xyl576807077** [mtreviso/linear-chain-crf#1]

**Comment by mtreviso** [comment]
Take a look here: https://github.com/mtreviso/linear-chain-crf/blob/master/crf_vectorized.py ;-)

**Comment by xyl576807077** [comment]
In _viterbi_decode function,  It needs to callback _find_best_path for every sample. So It doesn't mean to decode in parallel
Last night, I try to reduce a loop. You can refer to it, maybe it will help

```
    def _viterbi_decode(self, emissions: torch.FloatTensor,
                        mask: torch.ByteTensor) -> List[List[int]]:
        # emissions: (seq_length, batch_size, num_tags)
        # mask: (seq_length, batch_size)
        assert emissions.dim() == 3 and mask.dim() == 2
        assert emissions.shape[:2] == mask.shape
        assert emissions.size(2) == self.num_tags
        assert mask[0].all()

        seq_length, batch_size = mask.shape

        # Start transition and first emission
        # shape: (batch_size, num_tags)
        score = self.start_transitions + emissions[0]
        history = []

        # score is a tensor of size (batch_size, num_tags) where for every batch,
        # value at column j stores the score of the best tag sequence so far that ends
        # with tag j
        # history saves where the best tags candidate transitioned from; this is used
        # when we trace back the best tag sequence

        # Viterbi algorithm recursive case: we compute the score of the best tag sequence
        # for every possible next tag
        for i in range(1, seq_length):
            # Broadcast viterbi score for every possible next tag
            # shape: (batch_size, num_tags, 1)
            broadcast_score = score.unsqueeze(2)

            # Broadcast emission score for every possible current tag
            # shape: (batch_size, 1, num_tags)
            broadcast_emission = emissions[i].unsqueeze(1)

            # Compute the score tensor of size (batch_size, num_tags, num_tags) where
            # for each sample, entry at row i and column j stores the score of the best
            # tag sequence so far that ends with transitioning from tag i to tag j and emitting
            # shape: (batch_size, num_tags, num_tags)
            next_score = broadcast_score + self.transitions + broadcast_emission

            # Find the maximum score over all possible current tag
            # shape: (batch_size, num_tags)
            next_score, indices = next_score.max(dim=1)

            # Set score to the next score if this timestep is valid (mask == 1)
            # and save the index that produces the next score
            # shape: (batch_size, num_tags)
            score = torch.where(mask[i].unsqueeze(1), next_score, score)
            history.append(indices)

        # End transition score
        # shape: (batch_size, num_tags)
        score += self.end_transitions
        
        # shape: (4, batch_size, feats)
        history = torch.stack(history, dim=0)

        # shape: (batch_size,)
        seq_ends = mask.long().sum(dim=0) - 1
        best_tag_list = score.argmax(dim=1)
        best_tag_list = torch.unsqueeze(best_tag_list, dim=1)
        # Because my sentence length is 4
        for hist in reversed(history[:4]):
            best_last_tag = hist[torch.arange(hist.size(0)).long(), best_tag_list[:, -1]]
            best_last_tag = torch.unsqueeze(best_last_tag, dim=1)
            best_tag_list = torch.cat((best_tag_list, best_last_tag), dim=1)

        idx = [i for i in range(best_tag_list.size(1)-1, -1, -1)]
        idx = torch.tensor(idx, dtype=torch.long, device=best_tag_list.device)
        best_tag_list = torch.index_select(best_tag_list, dim=1, index=idx)
        return best_tag_list
```

**Comment by mtreviso** [comment]
Hi xyl576807077, 

Turns out that not every sample has the same length, so we would be wasting unnecessary computation over the padding symbols (imagine if max_seq_length is really long). Nevertheless, here is the code for doing this:

```python
best_sequences = max_final_tags.clone().unsqueeze(-1)
best_tag = max_final_tags
arange_vector = torch.arange(batch_size)
for backpointers_t in reversed(backpointers):
    best_tag = backpointers_t[best_tag, arange_vector]
    best_sequences = torch.cat((best_tag.unsqueeze(-1), best_sequences), dim=1)
best_sequences = best_sequences.tolist()
```
Filtering out pad positions:
```python3
best_sequences = [x[:seq_len[i]-1] + [x[-1]] for i, x in enumerate(best_sequences)]
```

I hope it helps!

**Comment by xyl576807077** [comment]
Thanks a lot!
